### ISSUE TITLE
nimbleList definitions and instances created in sampler_NUTS setup code

### ISSUE THREAD

**Issue by danielturek** [nimble-dev/nimbleHMC#30]
Something is going on with compiling the (nested) nimbleLists inside `sampler_NUTS`.  Specifically, when you try to compile an MCMC which contains more than one instance of `sampler_NUTS`, it gives a compilation error "Not all elements in the nimbleList list for compileNimbleList are from the same nimbleFunctionDef".

```
library(nimble)

## source samplers.R

code <- nimbleCode({
    a ~ dnorm(0, 1)
    b ~ dnorm(0, 1)
})

Rmodel <- nimbleModel(code, buildDerivs = TRUE)

conf <- configureMCMC(Rmodel, nodes = NULL)

conf$addSampler('a', 'NUTS')    ## using 'NUTS_classic'
conf$addSampler('b', 'NUTS')    ## for both will work

Rmcmc <- buildMCMC(conf)

Cmodel <- compileNimble(Rmodel)

Cmcmc <- compileNimble(Rmcmc, project = Rmodel)    ## error shown below

## Compiling
##   [Note] This may take a minute.
##   [Note] Use 'showCompilerOutput = TRUE' to see C++ compilation details.
## Error: Not all elements in the nimbleList list for compileNimbleList are from the same nimbleFunctionDef.  The class names include:nfRefClass_45 nfRefClass_47
```

**Comment by danielturek** [comment]
@perrydv Commented:

I suspect it's because the setup code creates instances of those `nimbleList` definitions. This differs from the first `sampler_HMC` which created the `nimbleList` definitions in setup code but did not create instances of them.

The easiest solutions that come to mind would be to create the `nimbleList` definitions outside of the sampler `setup` code, or to avoid creating instances in the setup code. The latter would require passing at least `state_current` around (by ref) through various method calls, so it was kind of nice to have it as a member variable.  A possible but tricky third solution would be to create the `nimbleList` definition once (on first call of `sampler_NUTS`) and put it in `nimbleUserNamespace` and then access it on subsequent calls from there. That would be arguably outside the API and therefore a bit of an expert maneuver.

**Comment by danielturek** [comment]
Yes, right, that makes sense.

I considered the option of (upon first build of `sampler_NUTS`) putting the nimbleList definitions into the `nimbleUserNamespace`, but decided it would be safer to not fiddle with the intended use of the namespace.

I ended up pulling the `nimbleList` definitions out of the `sampler_NUTS` setup code (and renaming the exported `nimbleList` definitions as `treebranchNL_NUTS` and `stateNL_NUTS`).  I then found that in order for the compiler to be happy, (it appears) these definitions also had to exist in the environment of the sampler function itself.  So in the `setup` code, I also added the lines:

```
treebranchNL <- treebranchNL_NUTS
stateNL <- stateNL_NUTS
```

And everything else flowed from there.  The cost of this is 2 additional exports from the `nimbleHMC` package.  The benefits are not fiddling with the namespace, and still using the system of pass-state-by-reference within the sampler.

If anyone wants to revisit these design decisions at any point, they're welcomed to.

Otherwise, these changes are pushed to branch `sampler_NUTS`.
